zlib-rs/src/read_buf.rs

```diff
@@ -70,85 +38,54 @@ impl<'a> ReadBuf<'a> {
-        // safety: filled describes how far into the buffer that the
-        // user has filled with bytes, so it's been initialized.
-        unsafe { slice_assume_init(slice) }
-    }
-
-    /// Returns a mutable reference to the entire buffer, without ensuring that it has been fully
-    /// initialized.
-    ///
-    /// The elements between 0 and `self.len()` are filled, and those between 0 and
-    /// `self.initialized().len()` are initialized (and so can be converted to a `&mut [u8]`).
-    ///
-    /// The caller of this method must ensure that these invariants are upheld. For example, if the
-    /// caller initializes some of the uninitialized section of the buffer, it must call
-    /// [`assume_init`](Self::assume_init) with the number of bytes initialized.
-    ///
-    /// # Safety
-    ///
-    /// The caller must not de-initialize portions of the buffer that have already been initialized.
-    /// This includes any bytes in the region marked as uninitialized by `ReadBuf`.
-    #[inline]
-    pub unsafe fn inner_mut(&mut self) -> &mut [MaybeUninit<u8>] {
-        self.buf
-    }
-
-    /// Returns the number of bytes at the end of the slice that have not yet been filled.
-    #[inline]
-    fn remaining(&self) -> usize {
-        self.capacity() - self.filled
+        &self.buf[..self.filled]
     }
 
     /// Clears the buffer, resetting the filled region to empty.
     ///
     /// The number of initialized bytes is not changed, and the contents of the buffer are not modified.
     #[inline]
     pub fn clear(&mut self) {
+        self.buf.fill(0);
         self.filled = 0;
     }
 
-    fn push_symbol(&mut self, a: u8, b: u8, c: u8) {
-        assert!(
-            self.remaining() >= 3,
-            "read_buf is full ({} bytes)",
-            self.capacity()
-        );
-
-        self.buf[self.filled] = MaybeUninit::new(a);
-        self.buf[self.filled + 1] = MaybeUninit::new(b);
-        self.buf[self.filled + 2] = MaybeUninit::new(c);
+    #[inline(always)]
+    pub fn push_lit(&mut self, byte: u8) {
+        // NOTE: we rely on the buffer being zeroed here!
+        self.buf[self.filled + 2] = byte;
 
-        self.initialized = Ord::max(self.initialized, self.filled + 3);
         self.filled += 3;
     }
 
-    pub fn push_lit(&mut self, byte: u8) {
-        self.push_symbol(0, 0, byte)
-    }
-
+    #[inline(always)]
     pub fn push_dist(&mut self, dist: u16, len: u8) {
+        let buf = &mut self.buf[self.filled..][..3];
         let [dist1, dist2] = dist.to_le_bytes();
-        self.push_symbol(dist1, dist2, len)
+
+        buf[0] = dist1;
+        buf[1] = dist2;
+        buf[2] = len;
+
+        self.filled += 3;
     }
 
     pub(crate) fn new_in(alloc: &Allocator<'a>, len: usize) -> Option<Self> {
         let buf = alloc.allocate_slice::<u8>(len)?;
 
-        Some(Self {
-            buf,
-            filled: 0,
-            initialized: 0,
-        })
+        buf.fill(MaybeUninit::new(0));
```

Comment:
Allocators often have an option to directly allocate pre-zeroed memory. How hard would it be to make use of this when not being passed a custom allocator to use?

Reply:
it's possible, but a bit cursed maybe? you'd have to check that the allocator is one you know, and can then special-case to use `calloc` or similar.

Reply:
Or maybe you can add a field to `Allocator` and keep it either None or a function that forwards to the regular alloc function in the code that accepts an allocator from the C side and turns it into an `Allocator` instance. This way you did use the zeroed allocator method if the C side didn't specify an allocator to use.